Improve getEmailPendingQuery from EmailBundle performance

rwrz opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | staging
| Bug fix?                               | no
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | no
| Related user documentation PR URL      | 
| Related developer documentation PR URL | 
| Issue(s) addressed                     | Fixes https://github.com/mautic/mautic/issues/10394

#### Description:
I'm using Mautic with Amazon SES API and with a list bigger than 130k. Until the SES QUEUE is fixed and deployed, I notice that the manual sending was incredible slow (30-60seconds per manual request), and since SES force us a limit per second, I was dispatching 35 e-mails per request to avoid any issue. Well, my limit is 35 per second, but with this request taking more than 30 seconds, it didn't worth it. So I started to investigate and found the query problem and fixed.

Basically, changed the NOT EXISTS on the emails_stats to a NOT IN, this way MySQL store the subquery results and compare later, instead of a one comparison per line (at least, I think this was the issue). I tried several things and this change was the only one that fix the issue. From 30-60seconds, to ms... 
Now the entire routine, takes 2secs only.

Since I was fixing this, I added a few indexes needed to speedup this query a little more (didn't tested with them).

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Fill with a LOT of contacts (100k+)
3. Send at least 3 or 4 newsletters for all contacts (manually, batch of 35) - using amazon SES (not sure if needed)
4. The first newsletter will goes fast, but the second and third will slow down A LOT. The 4th will be a joke.

rwrz committed:
Improve getEmailPendingQuery from EmailBundle performance
- created index to all wheres
- changed from NOT EXISTS to NOT IN
- migrations
- tests
- dnc using not in instead of exists as well
- dnc tests

rwrz commented:
BTW, I don't know if this is the right way to add new indexes to the migrations. Let me know if I need to change something.

escopecz commented:
There are some tests failing. The code you are modifying was covered by tests so it's nicely visible what the changes in the generated query is in the failing test: https://github.com/mautic/mautic/pull/9406/checks?check_run_id=1398221715

Plead update the tests.

You'll have to add new migration for the indexes as well. Generate the migration file with `bin/console doctrine:migrations:migrate` Then run `bin/console doctrine:schema:update --dump-sql` to get the SQL queries for the indexes you've added to the entities. Then fill in the migration. An example: https://github.com/mautic/mautic/blob/staging/app/migrations/Version20200805185714.php

RCheesley commented:
@rwrz are you able to look at the failing tests please? The process for generating migrations is documented here: https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-6-migrations-needed

rwrz commented:
Sure. I will take a look as soon as possible. I'm more busy than usual those days.
But I still have to fix the "stats" calculation queries as well. They are incredible slow!! 15 or more minutes to open the e-mail view, then more 10 minutes to see the clicking stats, just ridiculous.

rwrz commented:
Fixed tests.
And added a migration, but I just created the migration following the template file, not doing the 

> `bin/console doctrine:migrations:migrate` Then run `bin/console doctrine:schema:update --dump-sql`

As suggested. Hope it won't a problem.

rwrz commented:
@RCheesley any news here?

adonaicruz reviewed (changes requested):
if you have to many DNC leads will be slower the query, the following path fixed it for me.
file: app\bundles\EmailBundle\Entity\EmailRepository.php
line 174 - `$dncQb->select('dnc.lead_id')`
line 262 - `->andWhere($q->expr()->notIn('l.id', $dncQb->getSQL()))`

rwrz commented:
@adonaicruz can you send a PR on my repository with those changes?

RCheesley commented:
@rwrz looks like this may need to be rebased against the Features branch, then we will need to get it tested for the upcoming 3.3 release in February!

rwrz commented:
@RCheesley just did and included @adonaicruz changes as well.
thank you!= )

RCheesley commented:
Great - pinging @mautic/core-team to review!

alanhartless commented:
This is basically reverting the change done by https://github.com/mautic/mautic/pull/6592. 

```
This changes the query that determines who is to receive an email from NOT INs to EXISTS/NOT EXISTS. For a database of 2 million records with a lot of DNC and stat records, the original query took over 15 minutes. This new query takes 53ms.
```

Are the new indexes going to provide the same results when it comes to millions of records in these tables?

rwrz commented:
Interesting. I didn't have the same results as #6592 is proposing, from 15min to 53ms... Currently, those queries are taking 15minutes using EXISTS/NOT EXISTS. 
About the indexes, even though the query was using it to improve, I didn't notice great improvements in regards to the indexes. Mostly was improved because of the "NOT IN" changes.

But, if it was changed to improve before and now we are rolling back to improve again, we may need to study deeper what is causing so much slowness in "some cases". It may need denormalization to really achieve good results on any DB configuration.

I had a Mautic with 140k users and was impossible to deliver a simple message in less than 24 hours before, after the query, it was taking 8hours... (which is still A LOT)
Unfortunately, I stopped to use Mautic because of this and several other issues.

kuzmany commented:
@rwrz can you try on same instance with revert PR https://github.com/mautic/mautic/pull/6592 and test what performance do you receive with code used before?

rwrz commented:
@kuzmany sorry, I don't have my installation anymore

dennisameling committed:
Merge branch 'features' into improve-mail-to-send-query

dadarya0 commented on app/migrations/Version20201221175210.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $shouldRunMigration = false; // Please modify to your needs
+
+        if (!$shouldRunMigration) {
```
This migration would never execute queries because it will be skipped due to `$shouldRunMigration` is false.

In prepUp, you should check that new added index are available or not if available then skip.

dadarya0 commented on app/migrations/Version20201221175210.php:
```diff
@@ -0,0 +1,45 @@
+
+    public function up(Schema $schema): void
+    {
+        $this->addSql('CREATE INDEX '.$this->prefix.'lead_donotcontact_idx_channel_lead_id ON '.$this->prefix.'lead_donotcontact (channel, lead_id)');
```
Index names are different in migration and entity.
In migration it is - lead_donotcontact_idx_channel_lead_id
In Entity it is - dnc_channel_lead_id_search
both should be same.

dadarya0 replied:
Same should be apply for other 2 indexes.

RCheesley commented:
Bumping from the release due to outstanding feedback from code review.

jstawski commented:
I have the same problem with viewing the email details: #10394, which seems to be related to the same issue, which is the query. I applied the indexes mentioned in this PR, but nothing has changed. Running the query directly into MariaDB is still slow. Yes, I was able to bring it down from 53 seconds to 25 seconds, but it is not because of the indexes, but because of the **IN** vs **EXISTS**.  So I agree with @rwrz and [his comment](https://github.com/mautic/mautic/pull/9406#issuecomment-790625814) that "It may need denormalization to really achieve good results on any DB configuration."

andersonjeccel commented:
Hey there! 👋

I noticed this PR has been open for quite a while now (wow, time flies! 🕰️) without any recent activity. Given the age of this PR and the changes that have happened in the project since then, it's probably best to close this one out.

If you still think this change would be valuable, feel free to open a fresh PR with an updated implementation. Otherwise, no worries – thanks for your interest in contributing!

I'll be closing this PR to help keep our project tidy. Thanks for understanding! 😊
